source/Calamari/Kubernetes/ResourceStatus/ResourceStatusChecker.cs

```diff
@@ -30,75 +41,144 @@ public class ResourceStatusChecker : IResourceStatusChecker
-                    .SelectMany(IterateResourceTree)
-                    .ToDictionary(resource => resource.Uid, resource => resource);
-
-                var newDeploymentStatus = GetDeploymentStatus(newDefinedResourceStatuses, definedResources);
-
-                reporter.ReportUpdatedResources(resourceStatuses, newResourceStatuses, ++checkCount);
-
-                resourceStatuses = newResourceStatuses;
-                definedResourceStatuses = newDefinedResourceStatuses;
-                deploymentStatus = newDeploymentStatus;
-                
-                timer.WaitForInterval();
-            }
-            while (!timer.HasCompleted() && deploymentStatus == DeploymentStatus.InProgress);
+                var result = new ResourceStatusTaskResult();
+                do
+                {
+                    ResourceIdentifier[] definedResources;
+                    bool stillWaitingForMoreResources;
+                    lock (resourceLock)
+                    {
+                        stillWaitingForMoreResources = waitingForMoreResources;
+                        definedResources = resources.ToArray();
+                    }
+
+                    if (definedResources.Any())
+                    {
+                        var definedResourceStatuses = resourceRetriever
+                                                         .GetAllOwnedResources(definedResources, kubectl, options)
+                                                         .ToArray();
+                        var resourceStatuses = definedResourceStatuses
+                                                  .SelectMany(IterateResourceTree)
+                                                  .ToDictionary(resource => resource.Uid, resource => resource);
+
+                        var deploymentStatus = stillWaitingForMoreResources
+                            ? DeploymentStatus.InProgress
+                            : GetDeploymentStatus(definedResourceStatuses, definedResources);
+
+                        reporter.ReportUpdatedResources(result.ResourceStatuses, resourceStatuses, ++checkCount);
+
+                        result = new ResourceStatusTaskResult(
+                            definedResources,
+                            definedResourceStatuses,
+                            resourceStatuses,
+                            deploymentStatus);
+                    }
+
+                    timer.WaitForInterval();
+                } while (!timer.HasCompleted() && result.DeploymentStatus == DeploymentStatus.InProgress);
+                isCheckingStatus = false;
+                return result;
+            });
+        }
+
+        public void AddResources(IEnumerable<ResourceIdentifier> newResources)
+        {
+            var newResourcesList = newResources.ToList();
+            log.Info($"Resource Status Check: {newResourcesList.Count} new resources have been added.");
+            AddResourcesInternal(newResourcesList);
+            timer?.Restart();
```

Comment:
Main thing to note out is that the way this works is that the timer will start when you start the `ResourceStatusReportExecutor` but it will be restarted everytime you add additional resources. So if you start the executor and take too long to add more it will time out. Unsure on exactly what that would look like but not sure if that's really a problem?

I could be worth trying to setup something that makes applying some yaml take a long time.. but I'm not sure if that's possible.

Reply:
Yeah fair enough - I wasn't sure if the timeout meant "after everything is applied" or "for the whole application process". Do you think it's worth discussing this in the channel with the team?

Reply:
Yeah, I think this is a problem that needs input from product and design